### PR TITLE
chore: support circular objects

### PR DESCRIPTION
**Pull request by mxschmitt**

### CODE REVIEW

File: src/Playwright.Tests/PageEvaluateHandleTests.cs [hunk 1]
```diff
@@ -100,15 +102,33 @@ public async Task ShouldAcceptMultipleNestedHandles()
...
+                var wasCorrectlySerialized = await Page.EvaluateAsync<bool>("a => a.b === a", a);
+                Assert.True(wasCorrectlySerialized);
+            }
+
+            {
+                // lists
+                var a = new List<object>();
+                a.Add(a);
+                var wasCorrectlySerialized = await Page.EvaluateAsync<bool>("a => window.a = a.length === 1 && a[0] && a[0] === a", a);
+                await Page.PauseAsync();
```
Comment (pavelfeldman) [inline src/Playwright.Tests/PageEvaluateHandleTests.cs:128]: Ditto

File: src/Playwright/Transport/Converters/ProtocolSerializer.cs [hunk 2]
```diff
@@ -115,54 +112,66 @@ public object Serialize(object value)
                 return new { r = new { p = regex.ToString(), f = regex.Options.GetInlineFlags() } };
             }
 
-            if (value is IDictionary dictionary && dictionary.Keys.OfType<string>().Any())
+            if (value is ExpandoObject)
             {
-                _visited.Add(value);
+                var o = new List<object>();
+                id = ++visitorInfo.LastId;
+                visitorInfo.Visited.Add(value, id);
```
Comment (pavelfeldman) [inline src/Playwright/Transport/Converters/ProtocolSerializer.cs:119]: Visited is a Dictionary (hash table), and you are using a complex (potentially recursive) value as a key. In Java that would stack overflow unless identity hash code and comparison are used, so you'll need something like https://stackoverflow.com/questions/8946790/how-to-use-an-objects-identity-as-key-for-dictionaryk-v for all your Dictionaries.

File: src/Playwright/Transport/Converters/EvaluateArgumentValueConverter.cs [hunk 3]
```diff
@@ -173,27 +182,37 @@ internal static object ParseEvaluateResult(JsonElement result, Type t)
...
         {
             if (parsed == null)
             {
                 return null;
             }
 
+            if (visited.ContainsKey(parsed))
+            {
+                return visited[parsed];
+            }
```
Comment (campersau) [inline src/Playwright/Transport/Converters/EvaluateArgumentValueConverter.cs:207]:
```suggestion
            if (visited.TryGetValue(parsed, out var value))
            {
                return value;
            }
```

File: src/Playwright/Transport/Converters/EvaluateArgumentValueConverter.cs [hunk 4]
```diff
@@ -37,22 +37,19 @@
...
-            if (_visited.Contains(value))
+            int id;
+            if (value == null)
             {
-                throw new JsonException("Argument is a circular structure");
+                return new { v = "null" };
             }
 
-            if (value == null)
+            if (visitorInfo.Visited.ContainsKey(value.GetHashCode()))
```
Comment (campersau) [inline src/Playwright/Transport/Converters/EvaluateArgumentValueConverter.cs:50]:
```suggestion
            if (visitorInfo.Visited.TryGetValue(value.GetHashCode(), out var ref))
```

File: src/Playwright/Transport/Converters/EvaluateArgumentValueConverter.cs [hunk 5]
```diff
@@ -37,22 +37,19 @@
...
             {
-                return new { v = "null" };
+                return new Dictionary<string, object> { ["ref"] = visitorInfo.Visited[value.GetHashCode()] };
```
Comment (campersau) [inline src/Playwright/Transport/Converters/EvaluateArgumentValueConverter.cs:52]:
```suggestion
                return new Dictionary<string, object> { ["ref"] = ref };
```

**Review by campersau — COMMENTED** [review]
Nit: Avoid multiple dictionary lookups.

File: src/Playwright/Transport/Converters/EvaluateArgumentValueConverter.cs [hunk 6]
```diff
@@ -311,13 +332,21 @@ private static object ParseEvaluateResultToExpando(JsonElement result)
...
+                    list.Add(ParseEvaluateResultToExpando(item, refs));
                 }
                 return list.ToArray();
             }
             return null;
         }
+
+        internal class VisitorInfo
+        {
+            public Dictionary<object, int> Visited { get; set; }
```
Comment (pavelfeldman) [inline src/Playwright/Transport/Converters/EvaluateArgumentValueConverter.cs:347]: This still suggests hashcode computation, we need an identity hash map instead.

**Review by pavelfeldman — APPROVED** [review]

**Pull request merged**